runtime/tests/interpreter/transfer_test.go

```diff
@@ -34,52 +34,139 @@ func TestInterpretTransferCheck(t *testing.T) {
 
 	t.Parallel()
 
-	ty := &sema.CompositeType{
-		Location:   utils.TestLocation,
-		Identifier: "Fruit",
-		Kind:       common.CompositeKindStructure,
-	}
-
-	valueDeclarations := stdlib.StandardLibraryValues{
-		{
-			Name: "fruit",
-			Type: ty,
-			// NOTE: not an instance of the type
-			ValueFactory: func(_ *interpreter.Interpreter) interpreter.Value {
-				return interpreter.NewStringValue("fruit")
+	t.Run("String value as composite", func(t *testing.T) {
+
+		t.Parallel()
+
+		ty := &sema.CompositeType{
+			Location:   utils.TestLocation,
+			Identifier: "Fruit",
+			Kind:       common.CompositeKindStructure,
+		}
+
+		valueDeclarations := stdlib.StandardLibraryValues{
+			{
+				Name: "fruit",
+				Type: ty,
+				// NOTE: not an instance of the type
+				ValueFactory: func(_ *interpreter.Interpreter) interpreter.Value {
+					return interpreter.NewStringValue("fruit")
+				},
+				Kind: common.DeclarationKindConstant,
 			},
-			Kind: common.DeclarationKindConstant,
-		},
-	}
-
-	typeDeclarations := stdlib.StandardLibraryTypes{
-		{
-			Name: ty.Identifier,
-			Type: ty,
-			Kind: common.DeclarationKindStructure,
-		},
-	}
-
-	inter, err := parseCheckAndInterpretWithOptions(t,
-		`
-          fun test() {
-            let alsoFruit: Fruit = fruit
-          }
-        `,
-		ParseCheckAndInterpretOptions{
-			CheckerOptions: []sema.Option{
-				sema.WithPredeclaredValues(valueDeclarations.ToSemaValueDeclarations()),
-				sema.WithPredeclaredTypes(typeDeclarations.ToTypeDeclarations()),
+		}
+
+		typeDeclarations := stdlib.StandardLibraryTypes{
+			{
+				Name: ty.Identifier,
+				Type: ty,
+				Kind: common.DeclarationKindStructure,
 			},
-			Options: []interpreter.Option{
-				interpreter.WithPredeclaredValues(valueDeclarations.ToInterpreterValueDeclarations()),
+		}
+
+		inter, err := parseCheckAndInterpretWithOptions(t,
+			`
+              fun test() {
+                  let alsoFruit: Fruit = fruit
+              }
+            `,
+			ParseCheckAndInterpretOptions{
+				CheckerOptions: []sema.Option{
+					sema.WithPredeclaredValues(valueDeclarations.ToSemaValueDeclarations()),
+					sema.WithPredeclaredTypes(typeDeclarations.ToTypeDeclarations()),
+				},
+				Options: []interpreter.Option{
+					interpreter.WithPredeclaredValues(valueDeclarations.ToInterpreterValueDeclarations()),
+				},
 			},
-		},
-	)
-	require.NoError(t, err)
+		)
+		require.NoError(t, err)
+
+		_, err = inter.Invoke("test")
+		require.Error(t, err)
+
+		require.ErrorAs(t, err, &interpreter.ValueTransferTypeError{})
+	})
+
+	t.Run("contract and restricted type", func(t *testing.T) {
```

Comment:
These tests passed the type checking (correctly) and only failed at run-time due to the defensive value transfer check, which in addition to the static check, also checks subtyping for the underlying value.